packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -96,6 +102,8 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
             // eslint-disable-next-line no-null/no-null
             runtime !== null ? runtime.logger : undefined, undefined, { sharedObjectId: uuid() });
 
+        this._serializer = new FluidSerializer(this.runtime.IFluidHandleContext);
```

Comment:
I don't really like creating and exposing a serilizer from the dds. i see how i may make some thing easier, but i think it makes the dss more complex, and more coupled a specific serializer implmentation. also, in your draft pr i saw that you had to extend the serializer interface to expose all the handles which was also not great. ideally the serializer interface stays small, it get's injected into the dds, and whoever injects it can pass in a serilizer that tracks handles or whatever, but that isn't exposed at the dds layer.

Reply:
i'm wondering if rather than a member of sharedObject, the serializer is just passed in on snapshot. that should give more flexiblity than a global serializer, but keeps the nice properties of injection

Reply:
@agarwal-navin yeah. looks like some comment got lost, but i think this thread covers it better anyway. basically, i don't like dds being coupled to a specific serializer implementation but understand how a global serializer is hard. hoping we can find a best of both

Reply:
At the moment we don't have anything like this but I imagine that in the future we'll have scenarios where DDSs will have DDS-specific serialization.  E.g. ink data structure may want to serialize to InkML or ISF.  So I think it's actually not bad for DDS to bring its own (potentially domain-specific) serializer.

Reply:
i also think you'll run into problems with ops being serialized polluting the handles in the serializer. this is problem either way

Reply:
i almost think the datastore runtime should handle serialization/deserialization for ops, and give the dss a fresh serializer for snapshots, but i'm not sure how feasible that is.

Reply:
that would keep dss handle agnostic, which is good for what i call out in my lost comments, using dss in other contexts, like without our runtimes

Reply:
@ChumpChief comments are acting wonky, just seeing your reply now. that's interesting, but i'm not sure how that would work at all with @navin-agawal is trying to do with garbage collection. i see what you describe as outputting an alternate format which is orthogonal the summary tree.

Reply:
Yeah I don't have strong feedback on this topic for this particular PR, just pointing out that the DDS probably needs to bring its own serializer implementation to enable that sort of scenario in the future, so I don't really feel bad about creating the serializer in the DDS even though it feels like decentralization.

Reply:
> the serializer is just passed in on snapshot

This is nice from an API perspective: the user is handed what they need when the need it.

> the datastore runtime should handle serialization/deserialization for ops, and give the dss a fresh serializer for snapshots

Related: SparseMatrix would like to serialize permutation segments differently for ops vs. snapshots.
```
// HACK: We need to include the allocated handle in the segment's JSON representation
//       for snapshots, but need to ignore the remote client's handle allocations when
//       processing remote ops.
```
One way to do this would be to change ISegment to accept a serializer (instead of return an object), in which case ISegment.serialize() could examine the serializer to see if it's writing an op or a snapshot.

>  I imagine that in the future we'll have scenarios where DDSs will have DDS-specific serialization. E.g. ink data structure may want to serialize to InkML or ISF

I was thinking serialization should be driver-specific rather than DDS specific.

Our JSON over-the-wire comes up as a red flag w/customers exploring agents which want to read/write to Fluid DDSes at computer vs. human speeds (most recently, it showed up in a conversation w/Jupyter [here](https://github.com/jupyterlab/rtc/issues/80#issuecomment-702095857)).

It would be lovely if the driver could choose it's over-the-wire format and provide an appropriate IFluidSerializer that encodes directly to JSON, BSON, MsgPack, Avro, Protobuf dynamic messages, whatever.  -- related to #2627.

(This is another case where I like Tony's suggestion of separate serializers, as a driver may choose a fast encoder for ops  and compact encoder for snapshots.)

Reply:
I was skeptical about passing a serializer to the snapshot API because a DDS might use the other serializer that it uses to serialize ops. However, looking at this discussion that seems like the better approach considering the fact that a DDS may want to serialize ops differently from snapshot. Or that a user may pass in a different serializer in the driver.

Basically, this is how I think it might work -
- We keep the current serializer in the data store runtime for serializing ops.
- For snapshot, the data store runtime passes in an IFluidSerializer to the DDS.
  - This IFluidSerializer is responsible for tracking the routes of any IFluidHandle that is serialized during snapshot.
  - It also exposes this list of routes via an API.
- SharedObject calls the above API to get the list of routes and stores it during snapshot.
- During garbage collection, an API is called on the DDS (getReferencedRoutes() or something similar) to return the above list of routes.
- If a user wants to use its own IFluidSerializer, it is responsible for tracking the list of routes during each snapshot.

Does this seem like a good solution that covers the cases discussed here?

Also, note that getReferencedRoutes() can be overridden by any DDS which can provide its own custom implementation for return the list of referenced routes.

Reply:
> This IFluidSerializer is responsible for tracking the routes of any IFluidHandle that is serialized during snapshot.

One thing to consider: This forces DDSs to keep de-serialized state in memory. I..e., I, as an author of DDS, can't build a map that serializes on set() and keeps on serialized state. I think that's Ok limitation, but we need to explicitly state it somewhere. That said, I think we would have a problem with summaries that include ops (like Sequence). They are already in serialized format, and extracting info about handles would require de-serialization and re-serialization, I think.

@DLehenbauer, I do not understand the driver reference here. Notion of handles does not exist outside of container runtime, driver and storage should not care or even know what is a handle. Plus making format of snapshots storage specific breaks FRS scenarios and some other scenarios with moving files across storages. Storage should do minimal transformations on top of snapshot for export/import functionality to work.

Reply:
Serialization happens in two parts:

1.  We walk the object graph, replacing handles with a JSON objects that contains the absolute path + a marker 'type' field (see 'recursivelyReplaceHandles()')

2.  We invoke JSON.stringify(..) on the resulting object.

It would be nice to have the flexibility to choose something other than JSON for step # 2, even if all we do with it is get some real-world measurements to convince ourselves that JSON + gzip is fine.

Reply:
Some more thoughts (after talking aloud with Arin & Navin):
1. We should clearly articulate requirements for interfaces and not mix-in any unrelated things or implementation details. The new requirement here (in current implementation) is that DDS exposes to data store runtime (and data store runtime to container runtime) a set of routes out of it (i.e. handle URIs).
2. Serialization of handles is channel controlled/specific activity. DDS is the only thing that exists on channel today, but in future we are thinking of adding data stores on channels - this needs to be taken into account (fun exercise: we can generalize further and say that container runtime is a single channel in container). From minimalistic protocol level, channel needs only a route to itself from root (container runtime) to correctly implement serialization and deserialization of handles, we should not put anything else on this interface (i.e. serialize is not part of protocol between channel and data store runtime).

Reply:
I like your principles.  Is there a rough idea of how reclamation would happen...  for example, is there a new backend service that periodically crawls the graph, or is it some unlucky client?

Reply:
TBD. We need to clearly define contracts here. I.e. there are couple stages / properties of content in storage:
1. Live, included in all workstreams.
2. "deleted", i.e. not used when doing search indexing, not included in snapshot (used to boot container quickly), but is still present in storage and can be "reclaimed" by clients if done in reasonable amount of time (let's say 2 weeks) - this is needed for offline clients, clients that have such content in undo/redo in memory, or copy/paste.
3. Truly deleted from storage. Done by server (and possible in addition to that - by client) when time horizon in # 2 is over..

But nothing yet set in stone, we need to more clearly describe all those states and transitions between them and also account for differences in storage implementations.

Reply:
@agarwal-navin can you summarize the discussion here and lay out the path forward, maybe in the issue. There is a lot being discussed here and i want to make sure we're all on the same page.

Reply:
I was playing around with a few different prototypes for this, particularly w.r.t. garbage collection. I will talk about them and which one I am leaning towards. But first, from the discussion in this thread, here are the key principles we want to maintain:
1. Keep the IFluidSerializer interface minimal and clean. All it needs is to be able to serialize and parse handles. We don't want to put anything GC specific in there (like exposing parsed handles), for instance.
2. DDS may want to serialize ops differently from snapshots.
3. We want to leverage the snapshot process and the data generated for garbage collection.
4. Having an IFluidSerializer per DDS for serializing everything does not seem like a good option.

With the above in mind, I tried the following:
1. Have each DDS implement a getReferencedRoutes() API. It goes through its data, finds out all the handles it has, gets the handle's path and returns this list.
    - Its a simple design and it works well. It is flexible such that a DDS is free to choose how it wants to track routes out of it. The disadvantages are that each DDS has to implement their own logic and we are not re-using the data generated during snapshot.
2. Have the snapshot() API return a list of routes out of it in addition to the snapshot tree. A DDS can implement its own logic to track the handles it serializes during snapshot. We can have a default implementation in SharedObject but a DDS can choose to write it own.
    - This is also simple and works fairly well. However, this approach lacks flexibility because it makes the snapshot and the GC process coupled together. A DDS has to return this information during snapshot even if it tracks it separately. It will also make it somewhat complex to separate the snapshot process from GC if we want to (although currently it makes sense for them to run together).
3. Have separate APIs for snapshot and for getting routes out of a DDS. During snapshot, pass in an IFluidSerializer that the DDS should use to serialize handles during the snapshot process. This IFluidSerializer is different from the one that is used to serialize ops (the DataStoreRuntime's IFluidSerializer) and can track all handles that it serializes (this is very similar to Tony suggestions in his comments)
    - This gives us the best of both the above approaches and incorporates the principles I talked about earlier:
    - It uses the data generated during snapshot for use in GC without coupling the two processes together.
    - It doesn't change the IFluidSerializer interface.
    - It uses different serializers for ops and for snapshots.
    - It is flexible such that a DDS can choose to track routes out of it separate from the snapshot process and return it via getReferencedRoutes().

I like the third approach and will update this PR with an implementation for that. Please let me know your comments / concerns.

Reply:
> Having an IFluidSerializer per DDS for serializing everything does not seem like a good option.

I'm not quite sure why you think so. 

It's hard to see what is the actual contract in # 3. What we want is to get snapshot & all routes used by given DDS. If we start passing serializer, then it's not clear that the intention is to have all routes/handles used by DDS. DDS may have virtualization support and may not actually serialize anything, but reuse existing blobs (that were not changed from last snapshot). It's not very clearly at interface level that the intention is to re-serialize all the content, because API doe snot speak that language.. This would be (at best) written in comments.

I'd prefer for contract to explicitly state what is needed - snapshot & set of routes. And SharedObject.ts can implement it by creating serialer object and passing it down to all concrete classes implementing DDSs.

Reply:
> I'm not quite sure why you think so.
I was mostly summarizing the points that Tony and Daniel made in this thread. I think passing the serializer to the SharedObject (DDS) adds flexibility and keeps the serializer implementation separate from the DDS. However, I don't have strong feelings towards this because I think it shouldn't be hard to change it as per requirement later.

I agree that we need to clearly define the contract. I like the idea of passing the serializer to the snapshot API and have the DDS use it to serialize all its handles that point to routes out of it. I guess you are leaning more towards # 2 in my previous comment where the DDS returns both the snapshot tree and routes out of it in the same API. I can see how it is more clear in that it states what it actually wants the API to do. My only concern being that it ties in the snapshot and GC process together but I think we can work with that.

Reply:
> My only concern being that it ties in the snapshot and GC process together but I think we can work with that.

Passing serializer (and using it for GC) has exactly same problem (that's hidden contract), but in addition to that it requires to force to use it for every handle, not just stuff I maybe incrementally snapshotting.

Reply:
As per the latest changes:
- Changed snapshot() API to return the following:
  - The snapshot tree.
  - The routes to fluid objects that are referenced. 
  - The path to the object that contains these routes. This is usually just the DDS id but it can be any path which the DDS understands.
- snapshot() does not accept a serializer but the SnapshotSerializer is now created inside each DDS instead of the SharedObject. I think it makes the implementation more clear and explicit. All the data that the API requires is generated and returned from the same place. Having it in the SharedObject centralized this process but it was a little confusing because the intention was not clear that the summarizer should be used to serialize all handles.
- Added unit tests to the DDSes to test that they return the referenced routes correctly for all handles.

Reply:
I've been thinking about how to do GC over bigger-than-memory DDSes like SharedLog (and the upcoming SharedTree), and think the following could work:

1.  DDSes produce trees of GC nodes, where blocks of previously serialized GC nodes can be referenced/reused.  (Maybe you can do that already with routes to internal GC nodes?)

2.  The GC process works by walking the graph, inserting nodes into a randomized [bloom filter](https://en.wikipedia.org/wiki/Bloom_filter), which builds a conservative set of marked nodes.

3.  If during the walk we encounter a node that intersects the bloom filter, we insert it into a set.  If we encounter a node already in the set, we've detected a cycle and stop descending.

After the walk terminates, we intersect the bloom filter with the total set of routes to see what can be safely deleted.  

Being conservative, the filter may falsely include some orphaned nodes due to collision (the probability of this depends on how we construct the filter).  However, because the filter changes on each GC pass, all orphaned nodes will eventually be collected.

Reply:
Closing this thread. Daniel created #4398 to track this for GC using minimal memory.